[RFC] Add more clint checks

ZyX-I opened the pull request:
Adds three more checks:
1. Macros line continuation character alignment is now rejected.
2. Indented preprocessor directives are rejected.
3. Cast operators with spaces are rejected.

Last thing is most subjectable: I would really like to do the opposite (reject cast operators _without_ spaces), but style guide suggests otherwise. Commits are done so that new linter errors fixed and fixes to the errors which popped out because of previous fixes are separate. Last check has (and will have) no change that fixes new errors because there are too much.

ZyX-I committed:
clint: Do not allow aligning line continuation characters

Rejected by https://neovim.io/develop/style-guide.xml#Horizontal_Whitespace.

Note: what to do with “empty” lines is not described in the style guide, neither
it is checked.

ZyX-I committed:
*: Fix errors from new linter checks

ZyX-I committed:
*: Also fix the adjacent errors

ZyX-I committed:
clint: Reject indented preprocessor directives

This is also not allowed by the style guide:
https://neovim.io/develop/style-guide.xml#Preprocessor_Directives.

ZyX-I committed:
*: Fix new linter errors

ZyX-I committed:
func_attr: Fix other new linter errors

brcolow commented on src/clint.py:
```diff
@@ -2499,6 +2502,27 @@ def CheckSpacing(filename, clean_lines, linenum, nesting_state, error):
 
+    if Search(r'\S {2,}\\$', line):
+        error(filename, linenum, 'whitespace/line_continuation', 5,
+              'Too much spaces before \\, line continuation character must be '
```
Too much spaces -> Too many spaces

ZyX-I committed:
clint: Add rules that forbids spaces after a cast operator

https://neovim.io/develop/style-guide.xml#Horizontal_Whitespace

Note that this errors out for e.g.

    if (has_mbyte) mb_copy_char((const char_u **)(&f), &t); \

(found in macros.h:151) or

    #define ECMD_SET_HELP   0x02    /* set b_help flag of (new) buffer before

(found in ex_cmds.h:11) (note `(new) buffer`). I left this as-is because these 
pieces of code are already caught by another rule (braces near if and `/*`-style 
comments). Other false positives I found were:

1. `FUNC_ATTR_NONNULL_ARG(1) FUNC_ATTR_WARN_UNUSED_RESULT`: rejected by 
   requiring type name to start with `[a-zA-Z_]` (this makes `1` not be 
   incorrectly recognized as a type).
2. `#define FOO(var) (var)`: rejected by creating `cast_line`.
3. `(expr) * (expr)`: rejected by constructing regexp so that unary operators 
   require no spaces after them.
4. `int f(void) FUNC_ATTR_PURE`: rejected by requiring line to start with 
   a space.
5. `"." STR(NVIM_VERSION_PATCH) NVIM_VERSION_PRERELEASE`: not rejected, such 
   thing should be rare and it would be easy to get rid of the false positive by 
   e.g. breaking line.

Struct literals like `(MyStruct) { 4, 2 }` are not checked:

1. I am not sure whether they are under this rule at all (this is not a cast).
2. It would be hard to get rid of false positives (like the example with `if` 
   above, but now for *valid* `if() {}`s).

justinmk commented:
No objections.
